### ISSUE TITLE
Select a different html template file while composer is open

### ISSUE THREAD

**Issue by jandis82** [RealRaven2000/SmartTemplate4#24]
Dear, I would like to ask you for a new feature as was used in the Stationery - changing templates in a opened e-mail window after right click mouse.
![obrázek 9](https://user-images.githubusercontent.com/58221911/69642066-7e983d80-1061-11ea-8559-29dd977d2f97.jpg)

**Comment by RealRaven2000** [comment]
I will look into it - the main problem is that SmartTemplate⁴ is designed from the ground up to destroy and rebuild the email, which makes it difficult / impossible to rescue anything you have written. It would probably be possible to implement it to select before you start writing.

As a workaround, if you want to follow up to a machine generated / web form email and "open as new" or use attachments, you can alternatively use the reply / forward commands and their file template dropdowns.

**Comment by RealRaven2000** [comment]
I have just worked on [issue 7](https://github.com/RealRaven2000/SmartTemplate4/issues/24) "Losing all text in compose window when changing identity / signature not updated" which shows some behavior that will also apply if the file template is changed.

the main problem with changing the file template is that it isn't possible to keep any text that the user may already have started typing within the (wrong) template. SmartTemplate⁴ simply has no way of distinguishing the "automatically generated" and the "typed" parts of the Email. If you consider a simple template like

```
Dear %from(firstname)%,
%cursor%
Yours, 
  %identity(firstname)%
```

and you start typing in the cursor position, now having something like:

Dear Axel,
**how was the meeting?**
Yours, 
  Jan

To SmartTemplate⁴ significance of the 4 lines is exactly identical, they are just content within the smarttemplate div. They are not part of the template anymore at this stage, just some arbitrary HTML. 

When the users requests a new Template, (which may contain similar or totally different lines) the only thing SmartTemplate⁴ can do is throw away these lines and start generating the email from scratch; which means you have to type "how was the meeting?" and everything else again. 

Is this what you had in mind?

**Comment by RealRaven2000** [comment]
Maybe a nice workaround would be for requiring the user to select the written stuff that we want to keep before selecting "Change Stationery..." to show to Thunderbird that this text should be inserted somewhere (maybe at %cursor% ?) when the Email is rebuilt from the newly selected template?

**Comment by GrantHorizons** [comment]
Stationery is useful, but I have used "QuickText" for some time - including the abandonment of the previous version after paying for it. :(
QuickText functionality in SmartTemplate seems to be the way to go - it's already almost there, and using two different extensions to do similar things is not very elegant.
I have 248 templates in QT, and in around 21 categories - which makes for a busy toolbar row of dropdowns! MANY are only used rarely, but they're long so it's useful to have them pre-done. I use maybe 20-30 daily.
SmartTemplate with it's more advanced functionality and active support would be an ideal candidate to implement QT into. 
I'd want the "shortcut keys" of QT too, as I know that saves a LOT of time. Depending on what I want to insert, a single keystroke/tab inserts any of 36 templates instantly. Nice!
From there, a drop-down menu to choose from, with sub-menus, would be great. If that menu could also be activated optionally by a %menu% for example built in to a template, that would be nice, although it's important to be able to insert a template anytime anywhere. I often use different signatures depending on who, subject, status etc of the subject. 
Hope that helps!

**Comment by RealRaven2000** [comment]
> 
> 
> Stationery is useful, but I have used "QuickText" for some time - including the abandonment of the previous version after paying for it. :(
> QuickText functionality in SmartTemplate seems to be the way to go - it's already almost there, and using two different extensions to do similar things is not very elegant.
> I have 248 templates in QT, and in around 21 categories - which makes for a busy toolbar row of dropdowns! MANY are only used rarely, but they're long so it's useful to have them pre-done. I use maybe 20-30 daily.
> SmartTemplate with it's more advanced functionality and active support would be an ideal candidate to implement QT into.
> I'd want the "shortcut keys" of QT too, as I know that saves a LOT of time. Depending on what I want to insert, a single keystroke/tab inserts any of 36 templates instantly. Nice!
> From there, a drop-down menu to choose from, with sub-menus, would be great. If that menu could also be activated optionally by a %menu% for example built in to a template, that would be nice, although it's important to be able to insert a template anytime anywhere. I often use different signatures depending on who, subject, status etc of the subject.
> Hope that helps!

Thanks I think this should be handled in a separate issue. This one is about selecting a different template for the complete Email (kind of like changing the letter paper). It's similar, but not the same - inserting a "smart snippet" would be a wonderful additional function, I have to think about an efficient way to build the User Interface. Let's discuss in a new issue: See #28

**Comment by humeier** [comment]
The function to change the HTML template in the opened Composer should be very helpful for many former Stationery users.
Many thanks in advance.

**Comment by RealRaven2000** [comment]
Here is the first test version. In order to test you need to configure the composer toolbar first and add the a button using  right-click / customize:
![image](https://user-images.githubusercontent.com/6314674/80314400-5c117880-87e9-11ea-9146-94eb672ffa99.png)

you will have to restart Composer when you add this button for the first time. When composing the next mail, it will be filled with the stationery files from the three menus (write / reply or forward) depending on which case you are using for composing the mail.

To change the template select a different item from the dropdown

![image](https://user-images.githubusercontent.com/6314674/80314515-d3dfa300-87e9-11ea-8ec3-b2cbba514986.png)

I will very likely remove the last function "Configure menu item..." from this particular menu because it adds too much complexity.

If you start typing before changing the template, SmartTemplate cannot rescue the inserted text into the new template, and will display a warning to this extent.
 

[smartTemplate-fx-2.10.2pre31.zip](https://github.com/RealRaven2000/SmartTemplate4/files/4536004/smartTemplate-fx-2.10.2pre31.zip)

as always to install, unpack the xpi file first, then drag into extensions manager.

**Comment by RealRaven2000** [comment]
New version with the configuration menu item removed and the function for selecting any file from disk added:

[smartTemplate-fx-2.10.2pre35.zip](https://github.com/RealRaven2000/SmartTemplate4/files/4536181/smartTemplate-fx-2.10.2pre35.zip)

to install, unpack the xpi file first, then drag that into extensions manager.

**Comment by humeier** [comment]
For me, the template change in Composer now works as desired.
Thanks a lot

Am 26.04.2020 um 20:22 schrieb RealRaven2000:

**Comment by RealRaven2000** [comment]
> For me, the template change in Composer now works as desired. Thanks a lot 

Great. here is a new version with partly localized UI (the German version is complete, other locales will have missing strings replaced with English). It also includes a warning when you try to replace the template after having typed, hope the warning is clear enough:

[smartTemplate-fx-2.10.2pre41.zip](https://github.com/RealRaven2000/SmartTemplate4/files/4542458/smartTemplate-fx-2.10.2pre41.zip)

**Comment by RealRaven2000** [comment]
For my international users, here is just an updated version with completed translations for the locales fr, fi, ru, sr, es-ES, da, ja, pl, pt-PR and sv-SE. Also I completed all missing strings in the Serbian translation, with the help of google translate. Other translations have missing strings replaced with the English original strings, just like on a release version,

I also updated the [version history](http://smarttemplate4.mozdev.org/version.html#2.10.2) with instructions on how to add the button.

[smartTemplate-fx-2.10.2pre42.zip](https://github.com/RealRaven2000/SmartTemplate4/files/4544841/smartTemplate-fx-2.10.2pre42.zip)
